src/main/scala/gwi/text_transform/TextTransformTask.scala

```diff
@@ -0,0 +1,83 @@
+package gwi.text_transform
+
+import gwi.{Status, TaskState}
+
+import java.io.{InputStream, OutputStream}
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.io.Source
+import scala.util.{Failure, Success, Using}
+
+abstract class StatefullTask[S, R](initialState: S) {
+  private var taskState = initialState
+  protected val interrupted = new AtomicBoolean(false)
+
+  def getTaskState(): S = this.synchronized {taskState}
+  def setTaskState(state: S): Unit = this.synchronized {taskState = state}
+  def doTask(): R
+  def interruptTask(): Unit = {
+    interrupted.set(true)
+  }
+  protected def onTaskInterrupted(): Unit
+}
+
+abstract class TextTransformTask extends StatefullTask[TaskState, Unit](TaskState.initial) {
+
+  private var lastUpdate: Long = 0
+  protected var curLineIndex: Long = 0
+  private var linesInCurrentInterval: Long = 0 // for progress calculation
+
+  def processLine(line: String, destination: OutputStream): Unit
+
+  override def doTask(): Unit = {
+    setTaskState(TaskState(status = Status.Running.toString, 0))
+    lastUpdate = System.nanoTime()
+
+    try {
+      Using.Manager { use =>
+        val in  = use(getInputStream())
+        val out = use(getOutputStream)
+
+        val it = Source.fromInputStream(in).getLines()
+        while(it.hasNext) {
```

Comment:
Shouldn't we check the interruption flag here? 
```
while(it.hasNext && !interrupted.get())
```

Anyway, I think that it would be better to use a library with cancelation support than reinvent the wheel. e.g. cats-effect, fs2, akka-stream...

Reply:
no it is not correct to add the interrupted flag to the while loop because the streams can be closing or initialising while interrupted. the interruption is handled in the finally clause 
yes i tried first to implement it with fs2 but i need more time to do this properly.

Reply:
You could also use `Thread.currentThread().isInterrupted()`

Reply:
I think that it will never stop the computation now...

Reply:
I'm afraid that I don't understand. You call `t.future.cancel()` but never check if the thread was interrupted via `Thread.currentThread().isInterrupted()`